### PR TITLE
Rethrown exception call stack tests

### PR DESCRIPTION
**Pull request by ateoi**
Added tests to verify the reported call stack for rethrown exceptions, as specified by dotnet/coreclr#15780.

1. **Exception rethrown in a method different than the method where it was originally thrown:** the stack trace contains both the location in the method where the exception was originally thrown, and the location where the method that threw the exception was called.

2. **Exception is thrown and later rethrown in the same method:** the stack trace only contains the location where the exception was originally thrown and does not include the location where the exception was rethrown.

### CODE REVIEW

File: src/System.Runtime/tests/System/ExceptionTests.cs [hunk 1]
```diff
@@ -57,6 +59,103 @@ public static void Exception_TargetSite_Aot()
 
             Assert.True(caught);
         }
+
+        [Fact]
```
Comment (danmosemsft) [inline src/System.Runtime/tests/System/ExceptionTests.cs]: You will need to add `[SkipOnTargetFramework ...` for .NET Framework, until we port this. Usually we note the issue that fixed it for core.

Comment (danmosemsft) [reply src/System.Runtime/tests/System/ExceptionTests.cs]: Or possibly use PlatformDetection.IsFullFramework to verify .NET Framework behavior, so when it is ported, this test automatically fails and reminds us to enable it

**Comment by ateoi** [comment]
@danmosemsft I've added the full framework check, but NETFX x86 Release build is still failing. This is the output log:

``xUnit.net Console Runner (64-bit Desktop .NET 4.0.30319.42000)
    System.Tests.ExceptionTests.ThrowStatementDoesNotResetExceptionStackLineSameMethod [FAIL]
      Assert.True() Failure
      Expected: True
      Actual:   False
      Stack Trace:
           at System.Tests.ExceptionTests.VerifyCallStack(Stack`1 expectedCallStack, String reportedCallStack)
           at System.Tests.ExceptionTests.ThrowStatementDoesNotResetExceptionStackLineSameMethod()
    System.Tests.ExceptionTests.ThrowStatementDoesNotResetExceptionStackLineOtherMethod [FAIL]
      Assert.True() Failure
      Expected: True
      Actual:   False
      Stack Trace:
           at System.Tests.ExceptionTests.VerifyCallStack(Stack`1 expectedCallStack, String reportedCallStack)
           at System.Tests.ExceptionTests.ThrowStatementDoesNotResetExceptionStackLineOtherMethod()``

The fail is in the Assert.True call, where I check if the stack frame line matches the regular expression. From the error call stacks in the output, it seems that the source line code is missing, so I am guessing the symbol files for the test assemblies are missing, and that will cause the regular expression check.

**Comment by danmosemsft** [comment]
@safern could you please check that when building and running tests for NETFX, the symbol files are correctly next to the binaries the tests load?

**Comment by safern** [comment]
> @safern could you please check that when building and running tests for NETFX, the symbol files are correctly next to the binaries the tests load?

Checking now.

**Comment by ateoi** [comment]
@danmosemsft On Linux x64, the test is failing when checking the line number in the stack frame. This is the output log:

`
Assert.Equal() Failure
Expected: 122
Actual:   123
`

There's a difference between Windows and Linux here. Looking at the test source code, the line number reported in the exception stack trace on Linux is wrong. It seems a bug to me.

**Comment by ateoi** [comment]
@danmosemsft OSX failed on System.Net.HttpListener.Tests, it doesn't seem to be related to this test.

`
System.Net.Internals.SocketExceptionFactory+ExtendedSocketException : Connection refused 127.0.0.1:1300
`

**Comment by safern** [comment]
@danmosemsft I just doubled checked and it looks like built dlls (test and its imported projects) symbols are correctly bin placed in the execution folder.

**Comment by danmosemsft** [comment]
@mikem8361 could there be an off by one in Linux line numbers?

**Comment by danmosemsft** [comment]
&mikem8361

**Comment by danmosemsft** [comment]
@mikem8361

**Comment by mikem8361** [comment]
Yes there could be an off by one bug. I'm have investigate. Is there a simple repro?

**Comment by danmosemsft** [comment]
@mikem8361 probably you can pull down this branch on Linux and run the test

**Comment by danmosemsft** [comment]
@ateoi you can tell whether they are public or private by attaching in windbg  and using "lmv" -- or I think by using "symchk" both come with the Windows SDK:

https://docs.microsoft.com/en-us/windows-hardware/drivers/debugger/debugger-download-tools 

But you're building this assembly yourself right? In that case they are full private symbols. AFAIK you have to use a special tool to "strip" symbols to make public ones. We may not even do that for .NET Framework symbols anymore (not sure)

**Comment by ateoi** [comment]
@danmosemsft Ok, I think I understood the problem with source file and lines in NETFX. Looking at the build logs, it is creating a portable PDB for the test assemblies. According to the documentation at https://github.com/dotnet/core/blob/master/Documentation/diagnostics/portable_pdb.md

> Applications targeting .NET Framework 4.7.1 or earlier: printing stack traces with mappings back to line numbers (such as in an ASP.NET error page). The name of methods is unaffected, only the source file names and line numbers are unsupported

If that is the case, I think it's better to skip running these tests on NETFX. What do you think?

**Comment by danmosemsft** [comment]
@safern @weshaggard is there a good reason we don't build windows pubs for Netfx tests? They should only build on windows right? Or at least we can change it for building on windows.

**Comment by safern** [comment]
I just took a quick look and in did.props we’re setting all the pdbs to portable no matter the target framework, we should make them full/optimized for netfx vertical depending depending also on the build configuration. 

https://github.com/dotnet/corefx/blob/master/dir.props#L174-L198

I can throw a PR for that later today when I land.

**Comment by weshaggard** [comment]
>I just took a quick look and in did.props we’re setting all the pdbs to portable no matter the target framework, we should make them full/optimized for netfx vertical depending depending also on the build configuration.

We can do that but be sure to not break building this configuration on linux as we still need to support that as well.

**Comment by mikem8361** [comment]
We also need to be carefully not to break generating portable PDBs for Windows on .NET Core (or like Wes said any .NET Core platform).  For desktop/NETFX, Windows PDBs are ok, but not for .NET Core (it will break our VSCode symbols story).

**Comment by safern** [comment]
What configuration do you mean? The test configurations?

**Comment by safern** [comment]
Yes, I will only change the PDB type when we’re building for NETFX.

**Comment by weshaggard** [comment]
>Yes, I will only change the PDB type when we’re building for NETFX.

When you are building for NETFX on windows. You cannot do windows pdb generation on linux but we still need to have the builds produced there, obviously the netfx tests cannot run on linux but we still need to be able to build them there.

**Comment by safern** [comment]
Ok, makes sense. I didn’t know we could build the netfx vertical/tests on other platforms. Thanks

**Comment by safern** [comment]
I opened: https://github.com/dotnet/corefx/pull/28562 and ran this tests with that change and now only one test fails that is comparing for the line number.

```
ExceptionTests - reported call stack:
     at System.Tests.ExceptionTests.ThrowAndRethrowSameMethod(Stack`1 callStack) in E:\repos\corefx\src\System.Runtime\
  tests\System\ExceptionTests.cs:line 94
     at System.Tests.ExceptionTests.ThrowStatementDoesNotResetExceptionStackLineSameMethod() in E:\repos\corefx\src\Sys
  tem.Runtime\tests\System\ExceptionTests.cs:line 72
  * ExceptionTests - reported call stack:
     at System.Tests.ExceptionTests.ThrowException(Stack`1 callStack) in E:\repos\corefx\src\System.Runtime\tests\Syste
  m\ExceptionTests.cs:line 141
     at System.Tests.ExceptionTests.ThrowAndRethrowOtherMethod(Stack`1 callStack) in E:\repos\corefx\src\System.Runtime
  \tests\System\ExceptionTests.cs:line 133
     at System.Tests.ExceptionTests.ThrowStatementDoesNotResetExceptionStackLineOtherMethod() in E:\repos\corefx\src\Sy
  stem.Runtime\tests\System\ExceptionTests.cs:line 107
      System.Tests.ExceptionTests.ThrowStatementDoesNotResetExceptionStackLineOtherMethod [FAIL]
        Assert.Equal() Failure
        Expected: 131
        Actual:   133
        Stack Trace:
          E:\repos\corefx\src\System.Runtime\tests\System\ExceptionTests.cs(160,0): at System.Tests.ExceptionTests.Veri
  fyCallStack(Stack`1 expectedCallStack, String reportedCallStack)
          E:\repos\corefx\src\System.Runtime\tests\System\ExceptionTests.cs(111,0): at System.Tests.ExceptionTests.Thro
  wStatementDoesNotResetExceptionStackLineOtherMethod()
```

But now we get line numbers and file names.

**Comment by safern** [comment]
```
14:14:23       System.Tests.ExceptionTests.ThrowStatementDoesNotResetExceptionStackLineSameMethod [FAIL]
14:14:23         Assert.Equal() Failure
14:14:23         Expected: 72
14:14:23         Actual:   73
14:14:23         Stack Trace:
14:14:23           D:\j\workspace\windows-TGrou---2a8f9c29\src\System.Runtime\tests\System\ExceptionTests.cs(153,0): at System.Tests.ExceptionTests.VerifyCallStack(Stack`1 expectedCallStack, String reportedCallStack)
14:14:23           D:\j\workspace\windows-TGrou---2a8f9c29\src\System.Runtime\tests\System\ExceptionTests.cs(77,0): at System.Tests.ExceptionTests.ThrowStatementDoesNotResetExceptionStackLineSameMethod()
14:14:23       System.Tests.ExceptionTests.ThrowStatementDoesNotResetExceptionStackLineOtherMethod [FAIL]
14:14:23         Assert.Equal() Failure
14:14:23         Expected: 131
14:14:23         Actual:   133
14:14:23         Stack Trace:
14:14:23           D:\j\workspace\windows-TGrou---2a8f9c29\src\System.Runtime\tests\System\ExceptionTests.cs(153,0): at System.Tests.ExceptionTests.VerifyCallStack(Stack`1 expectedCallStack, String reportedCallStack)
14:14:23           D:\j\workspace\windows-TGrou---2a8f9c29\src\System.Runtime\tests\System\ExceptionTests.cs(112,0): at System.Tests.ExceptionTests.ThrowStatementDoesNotResetExceptionStackLineOtherMethod()
14:14:26     Finished:    System.Net.Requests.Tests
```

this are the failures coming from the same test.

**Comment by ateoi** [comment]
@danmosemsft 

> do you have time to take a look at that last failure?

I will take a look later today.

**Comment by ateoi** [comment]
@danmosemsft @mikem8361 
I am analyzing the stack trace line off by one issue, and this is not a specific Linux build problem. This seems to be an issue that happens only on Release builds.

For this test, I am getting inconsistent exception stack traces for debug and release builds. On debug builds, I get the expected exception stack traces, but on release builds I get different ones. I've tagged the inconsistencies below with **[right]** and **[wrong]**.

I probably will need help to figure out what is going on here.

**Windows NETFX**

1. ThrowStatementDoesNotResetExceptionStackLineSameMethod test
- DEBUG call stack:
<pre><code>at System.Tests.ExceptionTests.ThrowAndRethrowSameMethod(Stack`1 callStack) in C:\Repos\corefx\src\System.Runtime\tests\System\ExceptionTests.cs:line 91
<b>[right]</b> at System.Tests.ExceptionTests.ThrowStatementDoesNotResetExceptionStackLineSameMethod() in C:\Repos\corefx\src\System.Runtime\tests\System\ExceptionTests.cs:line 71</code></pre>
- RELEASE call stack:
<pre><code>at System.Tests.ExceptionTests.ThrowAndRethrowSameMethod(Stack`1 callStack) in C:\Repos\corefx\src\System.Runtime\tests\System\ExceptionTests.cs:line 91
<b>[wrong]</b> at System.Tests.ExceptionTests.ThrowStatementDoesNotResetExceptionStackLineSameMethod() in C:\Repos\corefx\src\System.Runtime\tests\System\ExceptionTests.cs:line 72</code></pre>

2. ThrowStatementDoesNotResetExceptionStackLineOtherMethod test
- DEBUG call stack:
<pre><code>at System.Tests.ExceptionTests.ThrowException(Stack`1 callStack) in C:\Repos\corefx\src\System.Runtime\tests\System\ExceptionTests.cs:line 134
at System.Tests.ExceptionTests.ThrowAndRethrowOtherMethod(Stack`1 callStack) in C:\Repos\corefx\src\System.Runtime\tests\System\ExceptionTests.cs:line 127
<b>[right]</b> at System.Tests.ExceptionTests.ThrowStatementDoesNotResetExceptionStackLineOtherMethod() in C:\Repos\corefx\src\System.Runtime\tests\System\ExceptionTests.cs:line 103</code></pre>
- RELEASE call stack:
<pre><code>at System.Tests.ExceptionTests.ThrowException(Stack`1 callStack) in C:\Repos\corefx\src\System.Runtime\tests\System\ExceptionTests.cs:line 134
at System.Tests.ExceptionTests.ThrowAndRethrowOtherMethod(Stack`1 callStack) in C:\Repos\corefx\src\System.Runtime\tests\System\ExceptionTests.cs:line 127
<b>[wrong]</b> at System.Tests.ExceptionTests.ThrowStatementDoesNotResetExceptionStackLineOtherMethod() in C:\Repos\corefx\src\System.Runtime\tests\System\ExceptionTests.cs:line 104</code></pre>

**Linux CORE**
1. ThrowStatementDoesNotResetExceptionStackLineSameMethod test
- DEBUG call stack:
<pre><code>at System.Tests.ExceptionTests.ThrowAndRethrowSameMethod(Stack`1 callStack) in /home/ateoi/git/corefx/src/System.Runtime/tests/System/ExceptionTests.cs:line 85
<b>[right]</b> at System.Tests.ExceptionTests.ThrowStatementDoesNotResetExceptionStackLineSameMethod() in /home/ateoi/git/corefx/src/System.Runtime/tests/System/ExceptionTests.cs:line 71</code></pre>
- RELEASE call stack:
<pre><code>at System.Tests.ExceptionTests.ThrowAndRethrowSameMethod(Stack`1 callStack) in /home/ateoi/git/corefx/src/System.Runtime/tests/System/ExceptionTests.cs:line 85
<b>[wrong]</b> at System.Tests.ExceptionTests.ThrowStatementDoesNotResetExceptionStackLineSameMethod() in /home/ateoi/git/corefx/src/System.Runtime/tests/System/ExceptionTests.cs:line 72</code></pre>

2. ThrowStatementDoesNotResetExceptionStackLineOtherMethod test
- DEBUG call stack:
<pre><code>at System.Tests.ExceptionTests.ThrowException(Stack`1 callStack) in /home/ateoi/git/corefx/src/System.Runtime/tests/System/ExceptionTests.cs:line 134
<b>[right]</b> at System.Tests.ExceptionTests.ThrowAndRethrowOtherMethod(Stack`1 callStack) in /home/ateoi/git/corefx/src/System.Runtime/tests/System/ExceptionTests.cs:line 117
<b>[right]</b> at System.Tests.ExceptionTests.ThrowStatementDoesNotResetExceptionStackLineOtherMethod() in /home/ateoi/git/corefx/src/System.Runtime/tests/System/ExceptionTests.cs:line 103</code></pre>
- RELEASE call stack:
<pre><code>at System.Tests.ExceptionTests.ThrowException(Stack`1 callStack) in /home/ateoi/git/corefx/src/System.Runtime/tests/System/ExceptionTests.cs:line 134
<b>[wrong]</b> at System.Tests.ExceptionTests.ThrowAndRethrowOtherMethod(Stack`1 callStack) in /home/ateoi/git/corefx/src/System.Runtime/tests/System/ExceptionTests.cs:line 118
<b>[wrong]</b> at System.Tests.ExceptionTests.ThrowStatementDoesNotResetExceptionStackLineOtherMethod() in /home/ateoi/git/corefx/src/System.Runtime/tests/System/ExceptionTests.cs:line 104</code></pre>

**Comment by danmosemsft** [comment]
Assuming the of by one is stable, I suggest tweaking  the tests to accommodate it and getting this committed. And open an issue for @mikem8361 to investigate.

**Comment by ateoi** [comment]
> And open an issue for @mikem8361 to investigate.

Where should I open this issue? corefx or coreclr?

**Comment by danmosemsft** [comment]
@ateoi probably CoreCLR. thanks!

**Comment by danmosemsft** [comment]
Change looks good to merge when tests are green.

**Comment by ateoi** [comment]
> probably CoreCLR. thanks!

Issue opened: dotnet/coreclr#17419

**Comment by ahsonkhan** [comment]
Should we be getting expection tests logging on successful test runs?

I ran the System.Runtime.Tests locally and even though all the tests pass, I am getting the following output:
```text
 Discovering: System.Runtime.Tests
  Discovered:  System.Runtime.Tests
  Starting:    System.Runtime.Tests
  * ExceptionTests - reported call stack:
     at System.Tests.ExceptionTests.ThrowAndRethrowSameMethod(ValueTuple`3& rethrownExceptionStackFrame) in D:\GitHub\F
  ork\corefx\src\System.Runtime\tests\System\ExceptionTests.cs:line 84
     at System.Tests.ExceptionTests.ThrowStatementDoesNotResetExceptionStackLineSameMethod() in D:\GitHub\Fork\corefx\s
  rc\System.Runtime\tests\System\ExceptionTests.cs:line 70
  * ExceptionTests - reported call stack:
     at System.Tests.ExceptionTests.ThrowException() in D:\GitHub\Fork\corefx\src\System.Runtime\tests\System\Exception
  Tests.cs:line 128
     at System.Tests.ExceptionTests.ThrowAndRethrowOtherMethod(ValueTuple`3& rethrownExceptionStackFrame) in D:\GitHub\
  Fork\corefx\src\System.Runtime\tests\System\ExceptionTests.cs:line 115
     at System.Tests.ExceptionTests.ThrowStatementDoesNotResetExceptionStackLineOtherMethod() in D:\GitHub\Fork\corefx\
  src\System.Runtime\tests\System\ExceptionTests.cs:line 102
  Finished:    System.Runtime.Tests

  === TEST EXECUTION SUMMARY ===
     System.Runtime.Tests  Total: 15544, Errors: 0, Failed: 0, Skipped: 0, Time: 9.337s
```

**Pull request merged**